File: frontend/src/utils/nif.js
const LETRAS_DNI = "TRWAGMYFPDXBNJZSQVHLCKE";
const LETRAS_CONTROL_CIF = "JABCDEFGHI";
const CIF_CONTROL_LETRA = "PQRSNW";
const CIF_CONTROL_DIGITO = "ABEH";

const normalizar = (valor) => valor.toUpperCase().replace(/[\s-]/g, "");

function esDniValido(v) {
  const m = /^(\d{8})([A-Z])$/.exec(v);
  return !!m && LETRAS_DNI[parseInt(m[1], 10) % 23] === m[2];
}

function esNieValido(v) {
  const m = /^([XYZ])(\d{7})([A-Z])$/.exec(v);
  if (!m) return false;
  const numero = "XYZ".indexOf(m[1]) + m[2];
  return LETRAS_DNI[parseInt(numero, 10) % 23] === m[3];
}

function esCifValido(v) {
  const m = /^([ABCDEFGHJKLMNPQRSUVW])(\d{7})([0-9A-J])$/.exec(v);
  if (!m) return false;
  const [, tipo, digitos, control] = m;

  let suma = 0;
  digitos.split("").forEach((c, i) => {
    const d = parseInt(c, 10);
    if (i % 2 === 0) {
      const doble = d * 2;
      suma += doble > 9 ? doble - 9 : doble;
    } else {
      suma += d;
    }
  });
  const digitoControl = (10 - (suma % 10)) % 10;
  const letraControl = LETRAS_CONTROL_CIF[digitoControl];

  if (CIF_CONTROL_LETRA.includes(tipo) || "KLM".includes(tipo)) return control === letraControl;
  if (CIF_CONTROL_DIGITO.includes(tipo)) return control === String(digitoControl);
  return control === String(digitoControl) || control === letraControl;
}

export function esNifCifValido(valor) {
  const v = normalizar(valor);
  return esDniValido(v) || esNieValido(v) || esCifValido(v);
}

export function nifAviso(valor) {
  if (!valor || !valor.trim()) return null;
  if (esNifCifValido(valor)) return null;
  return "No parece un NIF/CIF/NIE válido. Revisa que no haya errores; si es un identificador extranjero, puedes ignorar este aviso.";
}

File: backend/app/services/nif.py
"""Validación del dígito de control de NIF, NIE y CIF españoles."""

import re

LETRAS_DNI = "TRWAGMYFPDXBNJZSQVHLCKE"
LETRAS_CONTROL_CIF = "JABCDEFGHI"
CIF_CONTROL_LETRA = "PQRSNW"
CIF_CONTROL_DIGITO = "ABEH"

_DNI = re.compile(r"^(\d{8})([A-Z])$")
_NIE = re.compile(r"^([XYZ])(\d{7})([A-Z])$")
_CIF = re.compile(r"^([ABCDEFGHJKLMNPQRSUVW])(\d{7})([0-9A-J])$")


def normalizar_nif(valor: str) -> str:
    """Mayúsculas, sin espacios, guiones ni puntos y sin el prefijo de país ES."""
    v = re.sub(r"[\s.\-]", "", valor.upper())
    if len(v) == 11 and v.startswith("ES"):
        v = v[2:]
    return v


def _dni_valido(v: str) -> bool:
    m = _DNI.match(v)
    return bool(m) and LETRAS_DNI[int(m.group(1)) % 23] == m.group(2)


def _nie_valido(v: str) -> bool:
    m = _NIE.match(v)
    if not m:
        return False
    numero = str("XYZ".index(m.group(1))) + m.group(2)
    return LETRAS_DNI[int(numero) % 23] == m.group(3)


def _cif_valido(v: str) -> bool:
    m = _CIF.match(v)
    if not m:
        return False
    tipo, digitos, control = m.groups()
    suma = 0
    for i, c in enumerate(digitos):
        d = int(c)
        if i % 2 == 0:
            doble = d * 2
            suma += doble - 9 if doble > 9 else doble
        else:
            suma += d
    digito = (10 - suma % 10) % 10
    letra = LETRAS_CONTROL_CIF[digito]

    if tipo in CIF_CONTROL_LETRA or tipo in "KLM":
        return control == letra
    if tipo in CIF_CONTROL_DIGITO:
        return control == str(digito)
    return control in (str(digito), letra)


def nif_cif_valido(valor: str) -> bool:
    v = normalizar_nif(valor)
    return _dni_valido(v) or _nie_valido(v) or _cif_valido(v)
